unikernel.ml

```diff
@@ -530,7 +586,187 @@ struct
+        match
+          Utils.Json.
+            ( get "current_password" xs,
+              get "new_password" xs,
+              get "confirm_password" xs )
+        with
+        | ( Some (`String current_password),
+            Some (`String new_password),
+            Some (`String confirm_password) ) -> (
+            let now = Ptime.v (P.now_d_ps ()) in
+            let new_password_hash =
+              User_model.hash_password ~password:new_password ~uuid:user.uuid
+            in
+            match
+              ( String.equal user.password
+                  (User_model.hash_password
+                     ~password:("\"" ^ current_password ^ "\"")
+                     ~uuid:user.uuid),
+                String.equal new_password_hash user.password )
+            with
+            | true, false -> (
+                match String.equal new_password confirm_password with
+                | true ->
+                    if User_model.password_validation new_password then (
+                      let updated_user =
+                        User_model.update_user user ~password:new_password_hash
+                          ~updated_at:now ()
+                      in
+                      Store.update_user !store updated_user >>= function
+                      | Ok store' ->
+                          store := store';
+                          Middleware.http_response reqd ~title:"OK"
+                            ~data:"Updated password successfully" `OK
+                      | Error (`Msg err) ->
+                          Logs.warn (fun m -> m "Storage error with %s" err);
+                          Middleware.http_response reqd ~title:"Error" ~data:err
+                            `Internal_server_error)
+                    else
+                      Middleware.http_response reqd ~title:"Error"
+                        ~data:"New password must be atleast 8 characters."
+                        `Internal_server_error
+                | false ->
+                    Middleware.http_response reqd ~title:"Error"
+                      ~data:"New password and confirm password do not match"
+                      `Bad_request)
+            | true, true ->
+                Middleware.http_response reqd ~title:"Error"
+                  ~data:
+                    "Your new password cannot be the same as the previous \
+                     password."
+                  `Bad_request
+            | _ ->
+                Middleware.http_response reqd ~title:"Error"
+                  ~data:"The current password entered is wrong." `Bad_request)
+        | _ ->
+            Middleware.http_response reqd ~title:"Error"
+              ~data:
+                (Fmt.str "Update password: Unexpected fields. Got %s"
+                   (Yojson.Basic.to_string (`Assoc xs)))
+              `Bad_request)
+    | _ ->
+        Middleware.http_response reqd ~title:"Error"
+          ~data:"Update password: expected a dictionary" `Bad_request
```

Comment:
I'd simplify the implementation:
- (a) check that new_password and confirm_password are the same (and conform to the password policy)
- (b) check that current_password is good
- (c) now save the new password

I wouldn't bother to restrict that you cannot use the old password as your new password. Because: Why bother?